### PR TITLE
feat: ridge and pca

### PR DESCRIPTION
**Pull request by robincourant**

### CODE REVIEW

File: feature_engineering.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import pandas as pd
+import numpy as np
+#import sys
```
Comment (robincourant) [inline feature_engineering.py]: Remove these 2 lines.

File: feature_engineering.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import pandas as pd
```
Comment (robincourant) [inline feature_engineering.py]: - First: native python modules (ie: all modules [here](https://docs.python.org/3/py-modindex.html)) then, newline, then, installed modules as `pandas` etc, then, newline and import custom module as `data_preparation`
- Sort subgroups in alphabetical order modules

File: feature_engineering.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import numpy as np
+#import sys
+#sys.path.insert(1,r"C:\Users\Yakoubi\regression_pipeline")
+from data_preparation import *
+from sklearn.linear_model import Ridge
+from sklearn.feature_selection import SelectFromModel
+
+def feature_selection_ridge(dataset:pd.DataFrame):
```
Comment (robincourant) [inline feature_engineering.py]: Your method should begin with a verb ex: `select_features_ridge`.

File: feature_engineering.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    """
+    Function to select the relevant features to build a model
```
Comment (robincourant) [inline feature_engineering.py]: - Documenting functions is done this way : https://stackoverflow.com/questions/3898572/what-is-the-standard-python-docstring-format, using the reST format
- In a docstring, a sentence begins with an upper case and ends with a point (a point is missing at the end)

File: feature_engineering.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    """
+    data_work = load_dataset(dataset)
+    prepare_data(data_work) # Perform data preparation before coming up with PCA
```
Comment (robincourant) [inline feature_engineering.py]: Comments should be inserted in the line above.

File: feature_engineering.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    nb_columns = len(data_work.columns)
+    
+    #splitting label column from input column
```
Comment (robincourant) [inline feature_engineering.py]: Add an upper case and a whitespace after `#`.

File: feature_engineering.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    label_column=data_work.iloc[:,-1] 
+    data_work=data_work.iloc[:,:-1]
+
+    #Using Ridge regression to select the relevant
```
Comment (robincourant) [inline feature_engineering.py]: Add a whitespace.

File: feature_engineering.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    sel_=SelectFromModel(Ridge())
```
Comment (robincourant) [inline feature_engineering.py]: - Add whitespaces between operators (here `sel_ = SelectFromModel(Ridge())`
- Give an explicit name to the variable `sel_`

File: pca_reduction.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import pandas as pd
```
Comment (robincourant) [inline pca_reduction.py]: This file should be merged with `feature_engineering`.

File: pca_reduction.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import numpy as np
+#import sys
+#sys.path.insert(1,r"C:\Users\Yakoubi\regression_pipeline")
+from data_preparation import *
+from sklearn.decomposition import PCA
+
+
+def reduce_dimensionality(dataset: pd.DataFrame):
```
Comment (robincourant) [inline pca_reduction.py]: Change this method name according to your method (ex: select_feature_pca`)

**Review by robincourant — COMMENTED** [review]
- Remove the two `.pyc` files (check `.gitignore` file)
- Resolve threads pointed out in the comments

File: pca_reduction.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    if(nb_columns > 10): # Depending on the number of features, the PCA algorithm will need more or less variance (expressed in percentage) in order to return adequate number of components.
+        pca = PCA(.9999)
+    else:
+        pca = PCA(.999)
+    label_column = data_work.iloc[:,nb_columns-1] # label column of dataset is located at the last position.
+    data_work = data_work.iloc[:,0:nb_columns-1]
+    data_transformed = pca.fit_transform(data_work)
+    data_transformed=pd.DataFrame(data=data_transformed,columns=[f'principal components {i}' for i in range (np.shape(data_transformed)[1])]) # Concatenating transformed columns into one dataframe.
+    data_transformed['label'] = label_column # appending column label to the transformed dataframe at the last position.
+    return data_transformed
```
Comment (robincourant) [inline pca_reduction.py]: Add a newline at the end of the file.

File: pca_reduction.py (hunk 10 above)
Comment (robincourant) [inline pca_reduction.py]: Add a return hint (ie: what type of object your method returns). Here you should add: `-> pd.DataFrame`.

File: feature_engineering.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
```
Comment (robincourant) [inline feature_engineering.py:1]: In your case you should have:
```
import numpy as np
import pandas as pd
from sklearn.feature_selection import SelectFromModel, SelectKBest,f_regression
from sklearn.linear_model import Ridge

from data_preparation import load_data, prepare_data
```

File: feature_engineering.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+from sklearn.feature_selection import SelectFromModel, SelectKBest,f_regression
+from sklearn.linear_model import Ridge
+
+
+
+from data_preparation import *
+
+def select_features_ridge(filename:str):
+    """Select the relevant features using Ridge Regression method.
```
Comment (robincourant) [inline feature_engineering.py:16]: You should have: 

```
def select_features_ridge(filename:str) -> pd.Index:
"""Select the relevant features using Ridge Regression method.

:param filename: path to the dataset's '.csv' file.
:return: the selected features.
"""
```

Correct the others.

File: feature_engineering.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    :return: the selected features in an Index type
+
+    """
+    data_work = load_dataset(dataset)
+    # Perform data preparation before selecting features
+    prepare_data(data_work)
+    nb_columns = len(data_work.columns)
+    
+    # Splitting label column from input column
+    label_column=data_work.iloc[:,-1]
```
Comment (robincourant) [inline feature_engineering.py]: Don't forget whitespaces around operators.

File: feature_engineering.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    data_work=data_work.iloc[:,:-1]
+
+    # Using Ridge regression to select the relevant features
+    selection= SelectFromModel(Ridge())
+    selection.fit(data_work,label_column)
+    selected_features = data_work.columns[(selection.get_support())]
+    return selected_features
+
+def select_features_correlation(filename:str,number_of_feature_to_keep:int):
+    """Select the relevant features by looking at correlation
```
Comment (robincourant) [inline feature_engineering.py]: Newline and point are missing.

**Review by iqdecay — CHANGES_REQUESTED** [review]
Git commits should be made according to this : https://hackernoon.com/git-it-together-some-tips-on-commit-etiquette-and-best-practices-for-junior-developers-1f147b8dfd56

**Pull request merged**